aws/lambda/pytorch-auto-revert/pytorch_auto_revert/signal.py

```diff
@@ -0,0 +1,363 @@
+        return SignalStatus.PENDING in self.statuses
+
+    @property
+    def has_success(self) -> bool:
+        return SignalStatus.SUCCESS in self.statuses
+
+    @property
+    def has_failure(self) -> bool:
+        return SignalStatus.FAILURE in self.statuses
+
+    def events_by_status(self, status: SignalStatus) -> List[SignalEvent]:
+        """Get all events with the specified status."""
+        return [event for event in self.events if event.status == status]
+
+    # make iterable
+    def __iter__(self):
+        return iter(self.events)
+
+
+@dataclass
+class PartitionedCommits:
+    """
+    Represents the result of partitioning commits based on an autorevert pattern.
+    """
+
+    def __init__(
+        self,
+        failed: List[SignalCommit],
+        unknown: List[SignalCommit],
+        successful: List[SignalCommit],
+    ):
+        self.failed = failed
+        self.unknown = unknown
+        self.successful = successful
+
+    def failure_events_count(self) -> int:
+        return sum(c.statuses.get(SignalStatus.FAILURE, 0) for c in self.failed)
+
+    def success_events_count(self) -> int:
+        return sum(c.statuses.get(SignalStatus.SUCCESS, 0) for c in self.successful)
+
+
+class InfraCheckResult(Enum):
+    """Outcome of infra check based on partitioned commits."""
+
+    CONFIRMED = "confirmed"  # failure bracketed by two successes (not infra)
+    PENDING = "pending"  # pending events could still form the sandwich
+    RESTART_SUCCESS = "restart_success"  # no success after any failure
+    RESTART_FAILURE = "restart_failure"  # no failure after any success
+
+
+class Signal:
+    """A refined, column-like view of raw CI data for pattern detection.
+
+    - key: stable identifier for the signal (e.g., normalized job/test name)
+    - workflow_name: source workflow this signal is derived from
+    - commits: newest → older list of SignalCommit objects for this signal
+    """
+
+    def __init__(self, key: str, workflow_name: str, commits: List[SignalCommit]):
+        self.key = key
+        self.workflow_name = workflow_name
+        # commits are ordered from newest to oldest
+        self.commits = commits
+
+    def detect_fixed(self) -> bool:
+        """
+        Find the first commit with any non‑pending event; if it contains a success, consider the signal recovered.
+        """
+        # find the first non-pending existing commit in the signal
+        for commit in self.commits:
+            if not all(event.is_pending for event in commit):
+                # If we found a commit with no pending jobs, check if it has any failed jobs
+                return any(event.is_success for event in commit)  # recovered
+        return False
+
+    def detect_flaky(self) -> bool:
+        """
+        Checks whether signal is flaky, i.e. has both successful and failed events for any commit.
+        Note: false result can mean that there is not enough data to determine flakiness.
```

Comment:
add a note that the logic here can't exclude infra breakages.

Red -> Red -> Green can mean that there was a infra issue and it got fixed. but the signal is not flaky.

Reply:
true flakiness should only be identified by the pattern A -> (B ->)+ A, being RED -> GREEN -> RED or GREEN -> RED -> GREEN